### ISSUE TITLE
fix all false-negatives about there is no single space between a type annotation and [] or .... for Google Style

### ISSUE THREAD

**Issue by HuGanghui** [checkstyle/checkstyle#8205]
According to the latest Google Style  https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html#s4.6.2-horizontal-whitespace, 

> Between a type annotation and [] or ....

So valid cases are:
`String...` `String[]`, `String @NotNull []`, `String @NotNull ...`
Invalid cases:
`String ...` `String []`, `String @NotNull[]`, `String @NotNull...`

As we removing ELIPSIS token from config in #6707 to avoid false-positives (unwanted violations)
We need to make new Check to address cases above and do not make any false-positives, as general Check `NoWhitespaceBefore` should not be over complicated with nuances of annotations before token.

$ cat Test.java

```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

public class Test {

  @Target(ElementType.TYPE_USE)
  @interface NonNull {
  }

  @NonNull int @NonNull[] @NonNull[] fiel1; // false-negative
  @NonNull int @NonNull [] @NonNull [] field2;
  //@NonNull int @NonNull ... field3; // non-compilable
  //@NonNull int @NonNull... field4; // non-compilable


  public void foo2(final char[] param) {
  }

  public void foo1(final char [] param) {
  }

  public void foo3(final char @NonNull[] param) { // false-negative
  }

  public void foo4(final char @NonNull [] param) {
  }

  void test4(String... param) { 
  }

  void test3(String ... param) { // violation on 8.32, no violation is #6707 fixes
  }

  void test2(String @NonNull... param) { // false-negative
  }

  void test1(String @NonNull ... param) { // violation on 8.32, no violation aft #6707
  }
}
```
$ java -jar /var/tmp/checkstyle-8.32-all.jar -c /google_checks.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:31:21: '...' is preceded with whitespace. [NoWhitespaceBefore]
[WARN] /var/tmp/Test.java:37:30: '...' is preceded with whitespace. [NoWhitespaceBefore]
Audit done.

**Comment by romani** [comment]
@HuGanghui , please report issue as defined in issue template. In title state a problem briefly, proposed solution should be inside description. 
in my comment I already did almost all required.

**Comment by HuGanghui** [comment]
As mentioned in https://github.com/checkstyle/checkstyle/issues/6707#issuecomment-602116682, and https://github.com/checkstyle/checkstyle/pull/7916#issuecomment-620637555, in order to fix all false-negatives, I think we need to create a new check——WhitespaceBefore.

**Comment by HuGanghui** [comment]
I am pleasure to work on it when it is approved.

**Comment by nmancus1** [comment]
If we close this via #9922 , we can remove the following comments:

```bash
➜  checkstyle git:(issue-3145-b) grep -R "#8205"
src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputNoWhitespaceBeforeAnnotations.java:    @NonNull int @NonNull[] @NonNull[] fiel1; // ok until #8205
src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputNoWhitespaceBeforeAnnotations.java:    public void foo3(final char @NonNull[] param) { // ok until #8205
src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputNoWhitespaceBeforeAnnotations.java:    void test1(String... param) { // ok until #8205
src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputNoWhitespaceBeforeAnnotations.java:    void test2(String ... param) { // ok until #8205
src/it/resources/com/google/checkstyle/test/chapter4formatting/rule462horizontalwhitespace/InputNoWhitespaceBeforeAnnotations.java:    void test3(String @NonNull... param) { // ok until #8205

```

**Comment by mohitsatr** [comment]
@romani Please approve. false-negatives are google-style related.

**Comment by romani** [comment]
Fix is merged

**Comment by romani** [comment]
There are still references to this issue in Inputs
https://github.com/checkstyle/checkstyle/pull/17510/files

**Comment by Youssefalaa7** [comment]
@romani i think this issue should be closed there are no references for this issue in the code now
